### ISSUE TITLE
XML.parse_html yield invalid memory access

### ISSUE THREAD

**Issue by andremrsantos** [crystal-lang/crystal#2752]
Hi all,

I was running some test on the API I am working on and found the following issue. When I run `XML.parse_html()` with a empty string the program crashes and reports `Invalid memory access`

```
Invalid memory access (signal 11) at address 0x8
[4309874299] *CallStack::print_backtrace:Int32 +107
[4309833463] __crystal_sigfault_handler +55
[140735701570858] _sigtramp +26
[4309990313] *XML::Error::new<Pointer(LibXML::Error)>:XML::Error +9
[4310062567] *XML::from_ptr<Pointer(LibXML::Doc)>:XML::Node +39
[4310062516] *XML::parse_html<String>:XML::Node +68
[4309799127] __crystal_main +2199
[4309842264] main +40
```

I understand this is a edge case, but the error should be reported more gracefully which I understand as one of the main goals of crystal and on of the reasons I have been loving the language. You can check the code I used to replicate the error [here](https://gist.github.com/andremrsantos/f447a1bb73f3d1ccbad9e400b60bdc41).

I hope this can contribute to the language.

**Comment by kirbyfan64** [comment]
My guess is that `LibXML.htmlReadIO` returns a null pointer since the document's empty. `from_ptr` therefore calls `Error.new(LibXML.getLastError)`. But, since there was no error, `getLastError` returns `NULL`, causing `Error.new` to crash when accessing `error.value.message`.

**Comment by fernandes** [comment]
@andremrsantos has you tested with latest / head crystal?

testing on 0.17.4 returns wrong number of arguments: https://carc.in/#/r/10rw (using a HEAD version of compiler gives me the pretty same output)

if you are still experiencing the problem, try to reproduce on with a minimal example on carc (probably without webmock) I couldn't using `""` and `nil`

thank you

**Comment by andremrsantos** [comment]
@fernandes I running the latest crystal version: Crystal 0.17.4 (2016-05-26)

I re-run again but the error persisted. See example bellow

```
crystal eval 'require "xml"; XML.parse_html("")'                                                               add_download_clients
Invalid memory access (signal 11) at address 0x8
[4412899067] *CallStack::print_backtrace:Int32 +107
[4412883911] __crystal_sigfault_handler +55
[140735701570858] _sigtramp +26
[4412948665] *XML::Error::new<Pointer(LibXML::Error)>:XML::Error +9
[4412949319] *XML::from_ptr<Pointer(LibXML::Doc)>:XML::Node +39
[4412949268] *XML::parse_html<String>:XML::Node +68
[4412880682] __crystal_main +1370
[4412891192] main +40
```

**Comment by kirbyfan64** [comment]
It definitely still happens:

https://carc.in/#/r/10s1

The "Exited with: 11" is the error code of a segmentation fault. So it's definitely reproducible!

**Comment by asterite** [comment]
The way to solve this is to check if the string is empty and then raise `Document is empty (XML::Error)`. If you pass an empty IO it does raise that error. It's an inconsistency in libxml...

**Comment by fernandes** [comment]
Using the fix I proposed on #2759 now I got the following output when if @andremrsantos command

```crystal
○ ./bin/crystal eval 'require "xml"; XML.parse_html("")'
Using compiled compiler at .build/crystal
Document is empty (XML::Error)
[4409405474] *CallStack::unwind:Array(Pointer(Void)) +82
[4409405377] *CallStack#initialize<CallStack>:Array(Pointer(Void)) +17
[4409405336] *CallStack::new:CallStack +40
[4409400345] *raise<XML::Error>:NoReturn +25
[4409461980] *XML::parse_html<String>:XML::Node +60
[4409388831] __crystal_main +1423
[4409399736] main +40
```
